programs/bpf_loader/src/lib.rs

```diff
@@ -397,95 +407,110 @@ fn process_loader_upgradeable_instruction(
+                    instruction_context.get_index_in_transaction(first_instruction_account + 7)?,
+                )?);
+            let predrain_buffer = invoke_context
+                .feature_set
+                .is_active(&reduce_required_deploy_balance::id());
 
             // Verify Program account
-
-            if UpgradeableLoaderState::Uninitialized != program.state()? {
-                ic_logger_msg!(log_collector, "Program account already initialized");
-                return Err(InstructionError::AccountAlreadyInitialized);
-            }
-            if program.data_len()? < UpgradeableLoaderState::program_len()? {
-                ic_logger_msg!(log_collector, "Program account too small");
-                return Err(InstructionError::AccountDataTooSmall);
-            }
-            if program.lamports()? < rent.minimum_balance(program.data_len()?) {
-                ic_logger_msg!(log_collector, "Program account not rent-exempt");
-                return Err(InstructionError::ExecutableAccountNotRentExempt);
-            }
-
-            let new_program_id = *program.unsigned_key();
+            let new_program_id = {
+                let program = instruction_context
+                    .try_borrow_account(transaction_context, first_instruction_account + 2)?;
+                if UpgradeableLoaderState::Uninitialized != program.get_state()? {
+                    ic_logger_msg!(log_collector, "Program account already initialized");
+                    return Err(InstructionError::AccountAlreadyInitialized);
+                }
+                if program.get_data().len() < UpgradeableLoaderState::program_len()? {
+                    ic_logger_msg!(log_collector, "Program account too small");
+                    return Err(InstructionError::AccountDataTooSmall);
+                }
+                if program.get_lamports() < rent.minimum_balance(program.get_data().len()) {
+                    ic_logger_msg!(log_collector, "Program account not rent-exempt");
+                    return Err(InstructionError::ExecutableAccountNotRentExempt);
+                }
+                *program.get_key()
+            };
 
             // Verify Buffer account
-
-            if let UpgradeableLoaderState::Buffer { authority_address } = buffer.state()? {
+            let mut buffer = instruction_context
+                .try_borrow_account(transaction_context, first_instruction_account + 3)?;
+            if let UpgradeableLoaderState::Buffer { authority_address } = buffer.get_state()? {
                 if authority_address != upgrade_authority_address {
                     ic_logger_msg!(log_collector, "Buffer and upgrade authority don't match");
                     return Err(InstructionError::IncorrectAuthority);
                 }
-                if upgrade_authority_signer {
+                if !instruction_context.is_signer(first_instruction_account + 7)? {
```

Comment:
This is another case where the account being present was checked much earlier before.  There are probably other cases in this function